tests/integration_tests/functional/test_balloon.py

```diff
@@ -0,0 +1,713 @@
+    """Build a test matrix using the kernel with the balloon driver."""
+    artifacts = ArtifactCollection(_test_images_s3_bucket())
+    # Testing matrix:
+    # - Guest kernel: Linux 4.14
+    # - Rootfs: Ubuntu 18.04
+    # - Microvm: 2vCPU with 256 MB RAM
+    # TODO: Multiple microvm sizes must be tested in the async pipeline.
+    microvm_artifacts = ArtifactSet(artifacts.microvms(keyword="2vcpu_256mb"))
+    kernel_artifacts = ArtifactSet(artifacts.kernels(
+        keyword="vmlinux-4.14"
+    ))
+    disk_artifacts = ArtifactSet(artifacts.disks(keyword="ubuntu"))
+
+    # Create a test context and add builder, logger, network.
+    test_context = TestContext()
+    test_context.custom = {
+        'builder': MicrovmBuilder(bin_cloner_path),
+        'network_config': network_config,
+        'logger': logger,
+        'snapshot_type': SnapshotType.FULL,
+        'seq_len': 5
+    }
+
+    # Create the test matrix.
+    return TestMatrix(
+        context=test_context,
+        artifact_sets=[
+            microvm_artifacts,
+            kernel_artifacts,
+            disk_artifacts
+        ]
+    )
+
+
+def create_vm(context, deflate=True, stats_interval=0):
+    """Create a microvm with a balloon device."""
+    logger = context.custom['logger']
+    vm_builder = context.custom['builder']
+    snapshot_type = context.custom['snapshot_type']
+    enable_diff_snapshots = snapshot_type == SnapshotType.DIFF
+
+    logger.info(
+        "Testing {} with microvm: \"{}\", kernel {}, disk {} "
+        .format(
+            snapshot_type,
+            context.microvm.name(),
+            context.kernel.name(),
+            context.disk.name()
+        )
+    )
+
+    # Create a rw copy artifact.
+    root_disk = context.disk.copy()
+    # Get ssh key from read-only artifact.
+    ssh_key = context.disk.ssh_key()
+    # Create a fresh microvm from aftifacts.
+    microvm = vm_builder.build(
+        kernel=context.kernel,
+        disks=[root_disk],
+        ssh_key=ssh_key,
+        config=context.microvm,
+        enable_diff_snapshots=enable_diff_snapshots
+    )
+
+    network_config = net_tools.UniqueIPv4Generator.instance()
+    _, host_ip, guest_ip = microvm.ssh_network_config(
+        network_config,
+        '1',
+        tapname="tap0"
+    )
+    logger.debug("Host IP: {}, Guest IP: {}".format(host_ip, guest_ip))
+
+    copy_fillmem_to_rootfs(root_disk.local_path())
+
+    # Add a memory balloon with stats enabled.
+    response = microvm.balloon.put(
+        amount_mb=0,
+        deflate_on_oom=deflate,
+        must_tell_host=False,
+        stats_polling_interval_s=stats_interval
+    )
+    assert microvm.api_session.is_status_no_content(response.status_code)
+
+    return microvm
+
+
+def copy_fillmem_to_rootfs(rootfs_path):
+    """Build and copy the 'memfill' program to the rootfs."""
+    subprocess.check_call("gcc ./host_tools/fillmem.c -o fillmem", shell=True)
```

Comment:
I think this belongs in the `bin_cloner_path` fixture

Reply:
Not sure we want to rely on fixtures for this. I would prefer to do it explicitly.